keps/3125-maximum-execution-time/README.md

```diff
@@ -0,0 +1,208 @@
+
+## Proposal
+
+A new Kueue specific label is added `kueue.x-k8s.io/max-exec-time-seconds` containing
+the maximum number of seconds the Job is expected to execute.
+
+The duration is passed to the Job's Workload.
+
+If a workload stays admitted for longer that it is expected it will be Deactivated. 
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Batch User, I want to have the ability to set a limit on how long my job can run.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+#### Workload Spec
+
+Add a new optional field to hold the 
+
+```go
+// WorkloadSpec defines the desired state of Workload
+// +kubebuilder:validation:XValidation:rule="has(self.priorityClassName) ? has(self.priority) : true", message="priority should not be nil when priorityClassName is set"
+type WorkloadSpec struct {
+	// ...
+
+	// MaximumExecutionTimeSeconds if provided, determines the maximum time, in seconds, the workload can be admitted  
+	// befrore it's automatically deactivated.
+	MaximumExecutionTimeSeconds int32 `json:"maximumExecutionTimeSeconds,omitempty"`
+}
+```
+
+#### Workload Status
+
+Add a optional field to hold the 
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+	// ...
+
+	// AccumulatedPastExecutionTimeSeconds holds the total duration the workload spent in Admitted state
+	// in the previous `Admit` - `Evict` cycles.
+	AccumulatedPastExexcutionTimeSecond int32 `json:"accumulatedPastExexcutionTimeSeconds,omitempty"`
+}
+
+```
+
+### Constants
+
+Define a new constant holding the new label name `kueue.x-k8s.io/max-exec-time-seconds`.
+
+### Controller
+
+#### Workload
+During reconcile:
+
+- When a workload transitions out of `Admitted` state, add the time spent in its `status.AccumulatedPastExecutionTimeSeconds`
+```go
+cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
+wl.Status.AccumulatedPastExecutionTimeSeconds += wl.Spec.MaximumExecutionTimeSeconds -
+    time.Now().Sub(cond.LastTransitionTime.Time).Seconds()
+```
+- When the workload is admitted and a maximum execution time is specified compute the remaining execution execution time as:
+```go
+if wl.Spec.MaximumExecutionTimeSeconds != nil {
+	if cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted); cond != nil && cond.Status == metav1.ConditionTrue {
+		remainingTime := wl.Spec.MaximumExecutionTimeSeconds -
+            wl.Status.AccumulatedPastExecutionTimeSeconds -
+            time.Now().Sub(cond.LastTransitionTime.Time).Seconds()
+	}
+}
+```
+  - If `remainingTime` > 0 , a reconcile request should be queued with a `remainingTime` delay.
+  - If `remainingTime` <= 0 , the workload should be deactivated `wl.spec.Active = *false`, the `wl.status,AccumulatedPastExecutionTimeSeconds` set to 0, and a relevant event recorded.
```

Comment:
Shouldn't we follow the current deactivation mechanism instead of directly deactivating it?
`DeactivationTarget` -> `.spec.active=false`. I would like to keep consistency between all deactivation mechanisms, avoiding struggling.

Reply:
Doing this is the follow up reconcile is not a problem. 

I also want to use the same mechanism via DeactivationTarget. 

The intention of introducing DeactivationTarget was to also use it for ProvReq retries, so if it wasn't done I consider it an omission. 
cc @PBundyra who is currently working on the changes to the retry mechanism.

Reply:
For AdmissionCheck, I was supposed to switch the DeactivationTarget mechanism by @PBundyra at https://github.com/kubernetes-sigs/kueue/issues/1353.

Reply:
done